### PR TITLE
SNOW-2314161: In-band telemetry core pipeline

### PR DESCRIPTION
**Pull request by sfc-gh-turbaszek**
## Summary
- Add OTel-based telemetry types (`TelemetryEventType`, `ErrorOrigin`) and serialization from OTel spans/metrics to Snowflake's in-band JSON format
- REST endpoint for `POST /telemetry/send` with auth token and error handling
- Custom `SpanExporter` and `PushMetricExporter` that serialize and ship telemetry via the Snowflake endpoint
- Add `metrics` feature to `opentelemetry` and `opentelemetry_sdk` deps; `experimental_metrics_custom_reader` as dev-dep

## Test plan
- [x] 17 unit tests (types, serialization, exporter behavior)
- [x] 20 integration tests (REST endpoint, exporter pipeline with wiremock)
- [ ] CI green

🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Stack: 1/3** → #2 `telemetry/otlp-debug` → #3 `telemetry/wrapper-identity`

### CODE REVIEW

**Comment by sfc-gh-turbaszek** [comment]
* **#796** <a href="https://app.graphite.com/github/pr/snowflakedb/universal-driver/796?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#795** <a href="https://app.graphite.com/github/pr/snowflakedb/universal-driver/795?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#794** <a href="https://app.graphite.com/github/pr/snowflakedb/universal-driver/794?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#793** <a href="https://app.graphite.com/github/pr/snowflakedb/universal-driver/793?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/snowflakedb/universal-driver/793?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: sf_core/src/telemetry/snowflake_exporter.rs [hunk 1]
```diff
@@ -0,0 +1,445 @@
...
+
+            let payload = serialization::spans_to_snowflake_payload(&batch);
+            send_with_token(&session, &payload).await
+        }
+    }
+
+    fn shutdown_with_timeout(&mut self, _timeout: Duration) -> OTelSdkResult {
+        Ok(())
+    }
+}
```
Comment (sfc-gh-turbaszek) [inline sf_core/src/telemetry/snowflake_exporter.rs:97]: **TOCTOU race in metric exporter: token checked twice, can become None between checks**

The `PushMetricExporter::export` implementation reads the token lock to check availability (lines 88-93), drops the guard, serializes the payload, then calls `send_with_token` which acquires the lock *again*. Between the two lock acquisitions, the token can be revoked (set to `None`), causing `send_with_token` to silently drop the telemetry after the serialization work has already been done.

This isn't a correctness bug (the telemetry is best-effort), but the early check is misleading: it creates an impression that we avoid wasted serialization, yet it doesn't actually guarantee that. Consider either:
1. Removing the early token check and just relying on `send_with_token`'s check (simpler, same effective behavior), or
2. Holding the read guard across both serialization and send (but this would hold the lock longer, which may not be desirable).

Option 1 is probably the right call -- simplify and let `send_with_token` handle it.

File: sf_core/src/telemetry/types.rs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+/// Data provided by wrappers for session_init.
+#[derive(Debug, Clone)]
+pub struct SessionInitData {
+    pub driver_name: String,
+    pub driver_version: String,
+    pub language_runtime: String,
+    pub language_version: String,
+    pub language_compiler: Option<String>,
+    //pub release_date: Option<String>,
+    //pub is_lts: Option<bool>,
```
Comment (sfc-gh-turbaszek) [inline sf_core/src/telemetry/types.rs]: **Commented-out fields should not be checked in**

These commented-out fields (`release_date`, `is_lts`) look like WIP placeholders. They should either be included with a TODO comment explaining when they'll be enabled, or removed entirely and tracked in a follow-up issue. Commented-out code tends to rot and creates confusion about whether these are planned, deferred, or abandoned.

File: sf_core/src/rest/snowflake/telemetry.rs [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+    CommunicationSnafu, InvalidSnowflakeResponseSnafu, RequestConstructionSnafu, RestError,
+    UrlJoinSnafu, apply_json_content_type, apply_query_headers, read_response_json,
+};
+
+const TELEMETRY_SEND_PATH: &str = "/telemetry/send";
+const TELEMETRY_TIMEOUT: Duration = Duration::from_secs(30);
+
+#[derive(Debug, serde::Deserialize)]
+struct TelemetryResponse {
+    success: bool,
```
Comment (sfc-gh-turbaszek) [inline sf_core/src/rest/snowflake/telemetry.rs:17]: **`TelemetryResponse` is missing `message` and `code` fields present in all other Snowflake response types**

Looking at `HeartbeatResponse`, `RefreshSessionResponse`, and `AuthenticatorRequestResponse`, they all include `message: Option<String>` and `code: Option<String>`. The `TelemetryResponse` only has `success: bool`.

When `success == false`, the warning log on line 59 provides no diagnostic information. Including `message` and `code` would allow a more useful log like:
```rust
tracing::warn!(
    "Telemetry endpoint returned success=false: code={}, message={}",
    parsed.code.as_deref().unwrap_or("none"),
    parsed.message.as_deref().unwrap_or("none"),
);
```

This follows the pattern used by `heartbeat.rs` and makes debugging production issues much easier.

File: sf_core/src/telemetry/serialization.rs [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+
+    json!({ "logs": logs })
+}
+
+macro_rules! collect_sum_data_points {
+    ($logs:expr, $metric_name:expr, $sum:expr, $ty:ty) => {{
+        let timestamp = system_time_to_epoch_millis($sum.time());
+        for dp in $sum.data_points() {
+            let mut message = serde_json::Map::new();
+            message.insert("type".to_string(), Value::String($metric_name.to_string()));
```
Comment (sfc-gh-turbaszek) [inline sf_core/src/telemetry/serialization.rs]: **Macro defined after first use -- consider moving above or extracting to a helper function**

The `collect_sum_data_points!` macro is defined at line 56 but used starting at line 37. Rust allows this because of the `use collect_sum_data_points;` at line 73, but it makes the code harder to follow since readers encounter the invocations before seeing the definition.

More importantly, consider whether a macro is the right tool here. Since the three arms differ only in the Rust type (`u64`, `i64`, `f64`) and the `data_points()` / `value()` methods return types that all support `json!()` serialization, you could potentially use a generic helper function with a trait bound instead. That would give better error messages, IDE support, and type checking compared to a macro.

File: sf_core/src/telemetry/serialization.rs [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+        }
+    }};
+}
+use collect_sum_data_points;
+
+fn otel_value_to_json(value: &opentelemetry::Value) -> Value {
+    use opentelemetry::Value as OtelValue;
+    match value {
+        OtelValue::Bool(b) => Value::Bool(*b),
+        OtelValue::I64(i) => json!(*i),
```
Comment (sfc-gh-turbaszek) [inline sf_core/src/telemetry/serialization.rs:102]: **`unwrap_or_default()` silently maps pre-epoch times to 0**

If `SystemTime` is before `UNIX_EPOCH` (e.g., due to clock skew or a test bug), `duration_since(UNIX_EPOCH)` returns `Err` and `unwrap_or_default()` yields `Duration::ZERO`, producing a timestamp of `0`. This could make telemetry data confusing to debug on the server side -- a timestamp of `"0"` looks like missing data rather than a clock issue.

Consider logging a warning when this happens so it's at least observable:
```rust
fn system_time_to_epoch_millis(time: SystemTime) -> u128 {
    match time.duration_since(UNIX_EPOCH) {
        Ok(d) => d.as_millis(),
        Err(_) => {
            tracing::warn!("SystemTime before UNIX_EPOCH, using 0");
            0
        }
    }
}
```

File: sf_core/src/telemetry/types.rs [hunk 6]
```diff
@@ -0,0 +1,52 @@
+/// Where the error originated.
```
Comment (sfc-gh-turbaszek) [inline sf_core/src/telemetry/types.rs:1]: **`SessionInitData` and `WrapperErrorData` types are defined but never used**

These types are publicly exported but have no consumers in this PR. The `ErrorOrigin` enum is similarly unused beyond its own test. If these are intended for the next PR in the stack (#2 `telemetry/otlp-debug` or #3 `telemetry/wrapper-identity`), that's understandable, but it would be helpful to add a brief doc comment like `/// Used by wrapper-identity layer (see PR #N)` so reviewers know they're not dead code.

Without any usage, it's hard to review whether the API surface is correct -- for example, should `SessionInitData` fields like `driver_name` be `&str` or `Cow<'_, str>` to avoid unnecessary allocations when the data comes from static strings?

File: sf_core/src/telemetry/serialization.rs [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+/// Format: `{"logs": [{"message": {...}, "timestamp": "..."}]}`
+pub fn spans_to_snowflake_payload(spans: &[SpanData]) -> Value {
+    let logs: Vec<Value> = spans.iter().map(span_to_log_entry).collect();
+    json!({ "logs": logs })
+}
+
+fn span_to_log_entry(span: &SpanData) -> Value {
+    let mut message = serde_json::Map::new();
+
+    message.insert("type".to_string(), Value::String(span.name.to_string()));
```
Comment (sfc-gh-turbaszek) [inline sf_core/src/telemetry/serialization.rs:18]: **Span attribute keys can silently overwrite each other**

If a span has an attribute with key `"type"`, it will overwrite the span name that was inserted as `"type"` on line 18. Similarly, if an exception event has attributes with the same keys as the span's own attributes, the event attributes will silently overwrite them (since both use `message.insert()` on the same map).

This could lead to data loss in telemetry. Consider either:
- Prefixing the span name key (e.g., `"event.type"` or `"telemetry.type"`) to avoid collisions
- Documenting this as a known constraint ("span attributes must not use the key 'type'")
- Using a different data structure that preserves all values

File: sf_core/src/telemetry/mod.rs [hunk 8]
```diff
@@ -0,0 +1,6 @@
+pub mod types;
+
+#[doc(hidden)]
+pub mod serialization;
```
Comment (sfc-gh-turbaszek) [inline sf_core/src/telemetry/mod.rs:5]: **`#[doc(hidden)]` on public modules: consider `pub(crate)` instead**

The `serialization` and `snowflake_exporter` modules are marked `#[doc(hidden)]` but remain fully `pub`. This means downstream code *can* still depend on them -- they just won't appear in docs. If the intent is to make these internal implementation details, `pub(crate)` would be a stronger guarantee that prevents external usage entirely. If they need to be `pub` for the integration tests, that's worth documenting with a comment.

**Review by sfc-gh-turbaszek — COMMENTED** [review]
## Review Summary

This is a well-structured PR that adds the core in-band telemetry pipeline. The separation into REST endpoint, serialization, and exporter layers is clean, and the test coverage (37 tests across unit and integration) is solid. The best-effort error handling philosophy is correct for telemetry. A few substantive items to address:

### Key issues (posted as inline comments)

1. **Attribute key collision in serialization** -- the `"type"` key used for the span name can be overwritten by span attributes with the same key, and exception event attributes can overwrite span attributes. This is a silent data loss risk.

2. **TelemetryResponse missing `message`/`code` fields** -- every other Snowflake response type includes these, and they're needed for meaningful `success=false` diagnostics.

3. **TOCTOU in metric exporter** -- the early token-availability check provides a false sense of optimization since the token can change between the two lock acquisitions. Simplify by removing it.

4. **`#[doc(hidden)]` vs `pub(crate)`** -- if these modules are internal, use visibility restrictions rather than just hiding docs.

### Minor items

5. Commented-out fields in `SessionInitData` should be removed or documented.
6. `SessionInitData`, `WrapperErrorData`, `ErrorOrigin` are unused in this PR -- link to the consuming PR in the stack.
7. `system_time_to_epoch_millis` silently maps pre-epoch times to 0.
8. The `collect_sum_data_points!` macro could potentially be a generic function for better IDE/error support.

### What looks good

- The `send_with_token` pattern that holds the read guard across the reveal+send is a nice way to preserve `SensitiveString` zeroization guarantees.
- Delta temporality choice for metrics is correct for this use case (avoiding unbounded accumulation).
- The REST endpoint follows the existing `heartbeat.rs` pattern closely, making it easy to review.
- Integration tests cover the important edge cases: token revocation, token refresh, connection errors, server errors.

File: sf_core/src/telemetry/snowflake_exporter.rs (hunk 1 above)
Comment (sfc-gh-turbaszek) [reply sf_core/src/telemetry/snowflake_exporter.rs:97]: Done. Removed the early token check entirely — `send_with_token` handles it.

File: sf_core/src/telemetry/types.rs (hunk 2 above)
Comment (sfc-gh-turbaszek) [reply sf_core/src/telemetry/types.rs]: Done. Removed the commented-out fields.

File: sf_core/src/rest/snowflake/telemetry.rs (hunk 3 above)
Comment (sfc-gh-turbaszek) [reply sf_core/src/rest/snowflake/telemetry.rs:17]: Done. Added `message` and `code` fields to `TelemetryResponse` and included them in the warning log, matching the `HeartbeatResponse` pattern.

File: sf_core/src/telemetry/serialization.rs (hunk 4 above)
Comment (sfc-gh-turbaszek) [reply sf_core/src/telemetry/serialization.rs]: Done. Moved the macro definition above `metrics_to_snowflake_payload` and removed the `use` statement. Kept it as a macro for now since the three arms differ in concrete types (`u64`, `i64`, `f64`) — a generic function would need a trait bound that the OTel SDK doesn't expose cleanly.

File: sf_core/src/telemetry/serialization.rs (hunk 5 above)
Comment (sfc-gh-turbaszek) [reply sf_core/src/telemetry/serialization.rs:102]: Done. Now logs a warning when `SystemTime` is before `UNIX_EPOCH` instead of silently returning 0.

File: sf_core/src/telemetry/types.rs (hunk 6 above)
Comment (sfc-gh-turbaszek) [reply sf_core/src/telemetry/types.rs:1]: Done. Added doc comments linking these types to the consuming PR #795. Also changed `WrapperErrorData.error_source` from `String` to `ErrorOrigin` for type safety.

File: sf_core/src/telemetry/serialization.rs (hunk 7 above)
Comment (sfc-gh-turbaszek) [reply sf_core/src/telemetry/serialization.rs:18]: Done. Two changes:
1. Span attributes with key `"type"` are now skipped (with a warning log) to protect the span name field.
2. Exception event attributes are prefixed with `"exception."` (unless they already have the prefix) to avoid overwriting span-level attributes.

File: sf_core/src/telemetry/mod.rs (hunk 8 above)
Comment (sfc-gh-turbaszek) [reply sf_core/src/telemetry/mod.rs:5]: Can't use `pub(crate)` here — integration tests live outside the crate (`sf_core/tests/`) and need access to these modules. Added a comment explaining why they're `pub` + `#[doc(hidden)]` instead.

File: sf_core/src/telemetry/snowflake_exporter.rs [hunk 9]
```diff
@@ -0,0 +1,445 @@
...
+
+    if let Err(e) = rest::send_telemetry(
+        &session.client,
+        &session.server_url,
+        &session.client_info,
+        tokens.session_token.reveal(),
+        payload,
+    )
+    .await
+    {
```
[the remark this answers is not in the document]
Comment (sfc-gh-turbaszek) [reply sf_core/src/telemetry/snowflake_exporter.rs]: Done. `send_with_token` now clones the `SensitiveString` token under a short-lived read guard and drops the guard before the network call, avoiding lock contention with session refresh.
